Introduce Primary Terms

bleskes opened the pull request:
Every shard group in Elasticsearch has a selected copy called a primary. When a primary shard fails a new primary would be selected from the existing replica copies. This PR introduces `primary terms` to track the number of times this has happened. This will allow us, as follow up work and among other things, to identify operations that come from old stale primaries. It is also the first step in road towards sequence numbers.

Relates to #10708

bleskes commented:
@brwe @jasontedor  care to take a look?

brwe commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -637,6 +662,9 @@ public boolean equals(Object o) {
         if (unassignedInfo != null ? !unassignedInfo.equals(that.unassignedInfo) : that.unassignedInfo != null) {
             return false;
         }
+        if (primaryTerm != that.primaryTerm) {
```
do we need a change in `hashCode()` too?

brwe commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -500,6 +518,10 @@ public boolean equals(Object o) {
             return false;
         }
+
+        if (Arrays.equals(primaryTerms, that.primaryTerms) == false) {
```
should we change `hashCode()` too?

jasontedor replied:
The `version` field from this change and the `customs` field from a previous change are not incorporated into `IndexMetaData#hashCode`.

bleskes replied:
Agreed on the version - forgot it.  Will add it (and the customs as well)

brwe commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -794,6 +825,34 @@ public Builder version(long version) {
+            this.primaryTerms = primaryTerms;
+        }
+
+        private void allocatePrimaryTerms() {
```
can we call this initializePrimaryTerms ? I find allocate confusing here

brwe commented:
Should the primary term also increase when we move a primary from one node to another?

brwe commented:
When I restart a node then primaryTerm of primaries is incremented by 2. Is this intended?

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -580,6 +605,7 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeLong(version);
             out.writeByte(state.id);
             Settings.writeSettingsToStream(settings, out);
+            out.writeIntArray(primaryTerms);
```
Since we expect these to be non-negative and "not large", I wonder if it'd be better to serialize these using a variable-length encoding? See this [PR](https://github.com/elastic/elasticsearch/pull/14087).

bleskes replied:
we can - (and yeah, reviewed your PR :) )

bleskes replied:
updated to writeVIntArray

bleskes commented:
> Should the primary term also increase when we move a primary from one node to another?

We can maybe later. My feeling is now that this not needed and as this is the "same" primary - it just moved.

> When I restart a node then primaryTerm of primaries is incremented by 2. Is this intended?

Good catch!! fixed and added some testing.

brwe commented on core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java:
```diff
@@ -66,13 +67,13 @@ public void buildTestAllocator() {
     @Test
-    public void testNoProcessPrimayNotAllcoatedBefore() {
-        ShardRouting shard = TestShardRouting.newShardRouting("test", 0, null, null, null, true, ShardRoutingState.UNASSIGNED, 0, new UnassignedInfo(UnassignedInfo.Reason.INDEX_CREATED, null));
+    public void testNoProcessPrimacyNotAllocatedBefore() {
```
typo?

bleskes replied:
yep.. fixed.

brwe commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -482,6 +496,10 @@ public boolean equals(Object o) {
 
         IndexMetaData that = (IndexMetaData) o;
 
+        if (version != that.version()) {
```
should we change hashCode() too?

brwe commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -849,6 +917,11 @@ public static void toXContent(IndexMetaData indexMetaData, XContentBuilder build
             }
             builder.endObject();
 
+            builder.startArray("primary_terms");
```
can we make "primary_terms" a constant somewhere?

bleskes replied:
that's how that method used to be (many more values here) - i'll change them all...

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -203,17 +204,19 @@ public static State fromString(String state) {
         this.index = index;
         this.version = version;
+        this.primaryTerms = primaryTerms;
+        assert primaryTerms.length == numberOfShards();
```
Are we really okay with all of this repeated parsing and boxing and unboxing? Am I missing an obvious reason why we are not parsing this setting (and `number_of_replicas`) exactly once?

bleskes replied:
> Are we really okay with all of this repeated parsing and boxing and unboxing

I'm +1 on changing it - but it's a different change. This is not how performant code though  :)

> Also, I think it's best if in general we not invoke instance methods in constructors 

+1 here too - but this is just an assertion - I felt it was worth it.

jasontedor replied:
> it's a different change

Fair. I'll take it another PR.

> This is not how performant code though :)

I'm not following?

bleskes replied:
how == nigh (at least in mind mind). We don't call the number of shards methods very frequently

jasontedor replied:
I'm more concerned with the additional garbage which does impact the entire system.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -794,6 +827,34 @@ public Builder version(long version) {
+            if (numberOfShards() < 0) {
+                throw new IllegalStateException("you must set the number of shards before setting/reading primary terms");
+            }
+            primaryTerms = new int[numberOfShards()];
```
See my previous [questions](https://github.com/elastic/elasticsearch/pull/14062/files#r42002404) about double parsing.

bleskes replied:
double checking - I presume you mean the boxing/unbox one?  (which I answered).

jasontedor replied:
Yes.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -908,6 +981,19 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
+                        }
+                        for (int i = 0; i < list.size(); i++) {
+                            builder.primaryTerm(i, list.get(i));
+                        }
```
Is the `list` completely unnecessary if we just keep track of an index in the `while` loop and repeatedly call `builder#primaryTerm` as in the `for` loop? That will remove an allocation (and then the JVM does not even need to do escape analysis).

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -908,6 +981,19 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
+                        for (int i = 0; i < list.size(); i++) {
+                            builder.primaryTerm(i, list.get(i));
+                        }
+                        builder.primaryTerms(list.toArray());
```
Is this unnecessary since the elements were already copied one-by-one?

bleskes replied:
yikes! this was supposed to replace the previous one. I'll remove...

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -318,6 +331,7 @@ public void readFromThin(StreamInput in) throws IOException {
         }
 
         primary = in.readBoolean();
+        primaryTerm = in.readInt();
```
Make this `readVInt` (if and only if the write is changed to `writeVInt`)?

bleskes replied:
+1

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -363,6 +377,7 @@ public void writeToThin(StreamOutput out) throws IOException {
         }
 
         out.writeBoolean(primary);
+        out.writeInt(primaryTerm);
```
Make this `writeVInt` (if and only if the read is changed to `readVInt`)?

brwe commented:
I left some nitpicking but in general I wonder this: shard version and primary term should always be same for all copies. We add primary term to index meta data but not the shard version. Also, we write the shard version when we persist the shard meta data but not the primary term. Why do we treat them differently?
I also wonder why we cannot just get the version of the shard and the primary term from the index metadata instead of adding this information to the shard routings? It might be less confusing to have a single source of truth for this information and we write the index meta data now in any case.

bleskes commented:
>  shard version and primary term should always be same for all copies.

Note exactly - versions are incremented with every shard routing change, of any shard (primary or not). the terms are only incremented on primary assignment en promotion.

> Why do we treat them differently?

The shard versions are currently used for selected a primary after full cluster restart. Terms will be used for indexing. Long term, once we serialize allocation ids, we can consider removing the versions.

I rebased against master and pushed another commit. @jasontedor,  @brwe can you take another look?

brwe commented on core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java:
```diff
@@ -79,13 +75,71 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
+                metaDataBuilder.put(indexMetaDataBuilder);
+            }
+        }
+        if (metaDataBuilder != null) {
```
Can we increase increment the meta data version here? This comments says only master should do that (and did before): https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java#L437 Will the version not increase a second time once we get here https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java#L443 for example if we updated index meta data with some settings and this triggered this reroute?

bleskes replied:
I think you are right. This will result in double increments of the version. Good catch.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -408,19 +425,27 @@ public boolean equals(Object o) {
     @Override
     public int hashCode() {
         int result = index.hashCode();
+        result = 31 * result + (int) (version ^ (version >>> 32));
```
There's a built-in for this `[Long#hashCode(long)](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#hashCode-long-)` since Java 8.

bleskes replied:
sure thing.

brwe commented:
> Note exactly - versions are incremented with every shard routing change, of any shard (primary or not). the terms are only incremented on primary assignment en promotion.

What I meant was that they are both always the same for each copy (although primary term and version can of course differ). Shard version is only in the ShardRoutings but shard term is in both and that seems redundant to me.
I was actually hoping we could move the shard term to IndexMetaData only and not store this information in the ShardRouting too because I found it cumbersome to figure out where shard versions are incremented before and now we do the same thing for primary terms. But after some digging I think this is not really easy to do.

However, this is not a problem with this pull request but more with how versioning of MetaData, IndexMetaData, shards etc. works now. I have no good idea how to make this easier to read but opened an issue here to discuss: https://github.com/elastic/elasticsearch/issues/14158
We can leave it now in this pull request as is.

s1monw commented:
I look at the PR and I wonder if we should introduce a dedicated class for this for several reasons:

 * documentation, I think this needs a lot of documentation once used what it is and what it's semantics are.
 * we can implement `ToXContent`, `Comparable` and `Writeable`
 * we can ensure it's always positive and never decreasing
 * we should also use a long just to be on the safe end :) (over paranoid simon)
 
WDYT?

bleskes commented:
pushed another commit with a fix for the double version increment issue @brwe found and some(what) beefed java docs.

@s1monw I gave it some more thought and I still think - at least as things stand now - that a wrapper class for the PrimaryTerm will add complexity instead of making things clearer. It will just be a wrapper around an int and would obscure simple operation behind a method. Since it's a gut feeling thing I've asked the group today and @jasontedor tends to agree. We do totally see the importance of documentation. I've beefed up what I could in the current PR and added an explicit docs todo on the seq no meta data issue. I suggest we proceed as is. This is the very first step in a longer journey - as soon as there is more complex logic around the primary term that needs a home we'll wrap it up in a class.

I also moved primary terms to be long. I made them int originally to address concerns people voiced about 16 bytes (term + counter) per doc but I agree we can review it later on and maybe just encode it differently.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -702,6 +735,42 @@ public Builder version(long version) {
+        }
+
+        private void primaryTerms(long[] primaryTerms) {
+            this.primaryTerms = primaryTerms;
```
Should this be a copy?

bleskes replied:
Sure, I'll a copy for safety (though it's called with freshly constructed arrays).

jasontedor replied:
Maybe I misread, but I think there's one place where it's not in [`IndexMetaDataDiff.apply`](https://github.com/elastic/elasticsearch/pull/14062/files#diff-b103f001f952a0e6f2a4540d71593dd6R531)?

bleskes replied:
You didn’t missread - the only thing is that the diffs are read of the network and are discarded. That one pulled me over the line to actually change it and copy the array.

> On 21 Oct 2015, at 18:15, Jason Tedor <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
> 
> > +        }
> > +
> > +        /**
> > +         * sets the primary term for the given shard.
> > +         * See {@link IndexMetaData#primaryTerm(int)} for more information.
> > +         */
> > +        public Builder primaryTerm(int shardId, long primaryTerm) {
> > +            if (primaryTerms == null) {
> > +                initializePrimaryTerms();
> > +            }
> > +            this.primaryTerms[shardId] = primaryTerm;
> > +            return this;
> > +        }
> > +
> > +        private void primaryTerms(long[] primaryTerms) {
> > +            this.primaryTerms = primaryTerms;
> 
> Maybe I misread, but I think there's one place where it's not in IndexMetaDataDiff.apply?
> 
> —
> Reply to this email directly or view it on GitHub.
>

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -653,6 +684,8 @@ public int hashCode() {
         result = 31 * result + (currentNodeId != null ? currentNodeId.hashCode() : 0);
         result = 31 * result + (relocatingNodeId != null ? relocatingNodeId.hashCode() : 0);
         result = 31 * result + (primary ? 1 : 0);
+        result = 31 * result + (int) (primaryTerm ^ (primaryTerm >>> 32));
```
Another place for [`Long.hashCode`](https://docs.oracle.com/javase/8/docs/api/java/lang/Long.html#hashCode-long-)?

bleskes replied:
yep.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java:
```diff
@@ -653,6 +684,8 @@ public int hashCode() {
+        result = 31 * result + (int) (primaryTerm ^ (primaryTerm >>> 32));
+        ;
         result = 31 * result + (state != null ? state.hashCode() : 0);
         result = 31 * result + (int) (version ^ (version >>> 32));
```
While not part of your change, here [too](https://github.com/elastic/elasticsearch/pull/14062/files#r42509768), for consistency?

jasontedor replied:
I opened #14213 to just replace this everywhere that it's currently employed.

bleskes replied:
I'll rebase.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -816,11 +911,21 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
+                                throw new IllegalStateException("found a non-numeric value under [" + Fields.PRIMARY_TERMS.underscore() + "]");
+                            }
+                        }
+                        builder.primaryTerms(list.toArray());
```
I think that the `LongArrayList` is unnecessary with:

```java
@@ -19,7 +19,6 @@
 
 package org.elasticsearch.cluster.metadata;
 
-import com.carrotsearch.hppc.LongArrayList;
 import com.carrotsearch.hppc.cursors.ObjectCursor;
 import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
 import org.elasticsearch.Version;
@@ -912,15 +911,14 @@ public class IndexMetaData implements Diffable<IndexMetaData>, FromXContentBuild
                             }
                         }
                     } else if (fieldEquals(Fields.PRIMARY_TERMS, currentFieldName)) {
-                        LongArrayList list = new LongArrayList();
+                        int shardId = 0;
                         while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
                             if (token == XContentParser.Token.VALUE_NUMBER) {
-                                list.add(parser.longValue());
+                                builder.primaryTerm(shardId++, parser.longValue());
                             } else {
                                 throw new IllegalStateException("found a non-numeric value under [" + Fields.PRIMARY_TERMS.underscore() + "]");
                             }
                         }
-                        builder.primaryTerms(list.toArray());
                     }
                 } else if (token.isValue()) {
                     if (fieldEquals(Fields.STATE, currentFieldName)) {
```

bleskes replied:
I did it this way in order not to depend on the json order. We may not know how many shards are there (to allocated the primary terms array) if the settings are parsed after the terms. I felt this is better then an auto growing primary terms array because in other cases you want to have exceptions.

jasontedor replied:
Okay, that makes sense.

jasontedor commented:
I left a few more minor comments. I have reservations about the conversion from `int` to `long` but I think we can continue to think about that as this work progresses. Otherwise, LGTM.

bleskes commented:
this is pushed to the feature/seq_no branch. Thanks @jasontedor @brwe and @s1monw for the reviews.
